File: packages/tentacles/Services/Interfaces/web_interface/api/trading.py
#  Drakkar-Software OctoBot-Interfaces
#  Copyright (c) Drakkar-Software, All rights reserved.
#
#  This library is free software; you can redistribute it and/or
#  modify it under the terms of the GNU Lesser General Public
#  License as published by the Free Software Foundation; either
#  version 3.0 of the License, or (at your option) any later version.
#
#  This library is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
#  Lesser General Public License for more details.
#
#  You should have received a copy of the GNU Lesser General Public
#  License along with this library.
import flask

import octobot_services.interfaces.util as interfaces_util
import tentacles.Services.Interfaces.web_interface.util as util
import tentacles.Services.Interfaces.web_interface.login as login
import tentacles.Services.Interfaces.web_interface.models as models


def register(blueprint):
    @blueprint.route("/orders", methods=['GET', 'POST'])
    @login.login_required_when_activated
    def orders():
        if flask.request.method == 'GET':
            return flask.jsonify(models.get_all_orders_data())
        elif flask.request.method == "POST":
            result = ""
            request_data = flask.request.get_json()
            action = flask.request.args.get("action")
            if action == "cancel_order":
                if interfaces_util.cancel_orders([request_data]):
                    result = "Order cancelled"
                else:
                    return util.get_rest_reply('Impossible to cancel order: order not found.', 500)
            elif action == "cancel_orders":
                removed_count = interfaces_util.cancel_orders(request_data)
                result = f"{removed_count} orders cancelled"
            return flask.jsonify(result)


    @blueprint.route("/trades", methods=['GET'])
    @login.login_required_when_activated
    def trades():
        return flask.jsonify(models.get_all_trades_data())


    @blueprint.route("/positions", methods=['GET', 'POST'])
    @login.login_required_when_activated
    def positions():
        if flask.request.method == 'GET':
            return flask.jsonify(models.get_all_positions_data())
        elif flask.request.method == "POST":
            result = ""
            request_data = flask.request.get_json()
            action = flask.request.args.get("action")
            if action == "close_position":
                if interfaces_util.close_positions([request_data]):
                    result = "Position closed"
                else:
                    return util.get_rest_reply('Impossible to close position: position already closed.', 500)
            return flask.jsonify(result)


    @blueprint.route("/refresh_portfolio", methods=['POST'])
    @login.login_required_when_activated
    def refresh_portfolio():
        try:
            interfaces_util.trigger_portfolios_refresh()
            return flask.jsonify("Portfolio(s) refreshed")
        except RuntimeError:
            return util.get_rest_reply("No portfolio to refresh", 500)


    @blueprint.route("/currency_list", methods=['GET'])
    @login.login_required_when_activated
    def currency_list():
        return flask.jsonify(models.get_all_symbols_list())


    @blueprint.route("/historical_portfolio_value", methods=['GET'])
    @login.login_required_when_activated
    def historical_portfolio_value():
        currency = flask.request.args.get("currency", "USDT")
        time_frame = flask.request.args.get("time_frame")
        from_timestamp = flask.request.args.get("from_timestamp")
        to_timestamp = flask.request.args.get("to_timestamp")
        exchange = flask.request.args.get("exchange")
        try:
            return flask.jsonify(models.get_portfolio_historical_values(currency, time_frame,
                                                                        from_timestamp, to_timestamp,
                                                                        exchange))
        except KeyError:
            return util.get_rest_reply("No exchange portfolio", 404)


    @blueprint.route("/pnl_history", methods=['GET'])
    @login.login_required_when_activated
    def pnl_history():
        exchange = flask.request.args.get("exchange")
        symbol = flask.request.args.get("symbol")
        quote = flask.request.args.get("quote")
        since = flask.request.args.get("since")
        scale = flask.request.args.get("scale", "")
        return flask.jsonify(
            models.get_pnl_history(
                exchange=exchange,
                quote=quote,
                symbol=symbol,
                since=since,
                scale=scale,
            )
        )


    @blueprint.route("/clear_orders_history", methods=['POST'])
    @login.login_required_when_activated
    def clear_orders_history():
        return util.get_rest_reply(models.clear_exchanges_orders_history())


    @blueprint.route("/clear_trades_history", methods=['POST'])
    @login.login_required_when_activated
    def clear_trades_history():
        return util.get_rest_reply(models.clear_exchanges_trades_history())


    @blueprint.route("/clear_portfolio_history", methods=['POST'])
    @login.login_required_when_activated
    def clear_portfolio_history():
        return flask.jsonify(models.clear_exchanges_portfolio_history())


    @blueprint.route("/clear_transactions_history", methods=['POST'])
    @login.login_required_when_activated
    def clear_transactions_history():
        return flask.jsonify(models.clear_exchanges_transactions_history())

File: packages/tentacles/Services/Interfaces/web_interface/static/js/components/trading.js
/*
 * Drakkar-Software OctoBot
 * Copyright (c) Drakkar-Software, All rights reserved.
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3.0 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library.
 */


$(document).ready(async () => {
const addOrRemoveWatchedSymbol = (event) => {
    const sourceElement = $(event.target);
    const symbol = sourceElement.attr("symbol");
    let action = "add";
    if(sourceElement.hasClass("fas")){
        action = "remove";
    }
    const request = {};
    request["action"]=action;
    request["symbol"]=symbol;
    const update_url = sourceElement.attr("update_url");
    send_and_interpret_bot_update(request, update_url, sourceElement, watched_symbols_success_callback, watched_symbols_error_callback)
}

const watched_symbols_success_callback = (updated_data, update_url, dom_root_element, msg, status) => {
    create_alert("success", msg, "");
    if(updated_data["action"] === "add"){
        dom_root_element.removeClass("far");
        dom_root_element.addClass("fas");
    }else{
        dom_root_element.removeClass("fas");
        dom_root_element.addClass("far");
    }
}

const watched_symbols_error_callback = (updated_data, update_url, dom_root_element, result, status, error) => {
    create_alert("error", result.responseText, "");
}

const update_pairs_colors = () => {
    $(".pair_status_card").each((_, jselement) => {
        const element = $(jselement);
        const first_eval = element.find(".status");
        const status = first_eval.attr("status");
        if(status.toLowerCase().includes("very long")){
            element.addClass("card-very-long");
        }else if(status.toLowerCase().includes("long")){
            element.addClass("card-long");
        }else if(status.toLowerCase().includes("very short")){
            element.addClass("card-very-short");
        }else if(status.toLowerCase().includes("short")){
            element.addClass("card-short");
        }
    })
}

const get_displayed_orders_desc = () => {
    const orderDescs = [];
    const cancelButtonIndex = 8;
    $("#orders-table").DataTable().rows({filter: 'applied'}).data().map((value) => {
        orderDescs.push(value[cancelButtonIndex]);
    });
    return orderDescs;
}

const handleClearButton = () => {
    $("#clear-trades-history-button").on("click", (event) => {
        if (confirm("Clear trades history ?") === false) {
            return false;
        }
        const url = $(event.currentTarget).data("url")
        const success = (updated_data, update_url, dom_root_element, msg, status) => {
            // reload page on success
            reload_trades(true);
            reload_pnl(true);
        }
        send_and_interpret_bot_update(null, url, null, success, generic_request_failure_callback)
    })
}

const handle_cancel_buttons = () => {
    $("#cancel_all_orders").click((e) => {
        const to_cancel_orders = get_displayed_orders_desc();
        $("#ordersCount").text(to_cancel_orders.length);
        cancel_after_confirm($('#CancelAllOrdersModal'), to_cancel_orders, $(e.currentTarget).attr(update_url_attr), true);
    });
}

const handle_close_buttons = () => {
    $("button[data-action=close_position]").each((_, jsElement) => {
        $(jsElement).click((e) => {
            const element = $(e.currentTarget);
            close_after_confirm($('#ClosePositionModal'), element.data("position_symbol"),
                element.data("position_side"), element.data("update-url"));
        });
    });
}

const cancel_after_confirm = (modalElement, data, update_url, disable_cancel_buttons=false) => {
    modalElement.modal("toggle");
    const confirmButton = modalElement.find(".btn-danger");
    confirmButton.off("click");
    modalElement.keypress((e) => {
        if(e.which === 13) {
            handle_confirm(modalElement, confirmButton, data, update_url, disable_cancel_buttons);
        }
    });
    confirmButton.click(() => {
        handle_confirm(modalElement, confirmButton, data, update_url, disable_cancel_buttons);
    });
}

const close_after_confirm = (modalElement, symbol, side, update_url) => {
    modalElement.modal("toggle");
    const confirmButton = modalElement.find(".btn-danger");
    confirmButton.off("click");
    const data = {
        symbol: symbol,
        side: side,
    }
    modalElement.keypress((e) => {
        if(e.which === 13) {
            handle_close_confirm(modalElement, confirmButton, data, update_url);
        }
    });
    confirmButton.click(() => {
        handle_close_confirm(modalElement, confirmButton, data, update_url);
    });
}

const handle_close_confirm = (modalElement, confirmButton, data, update_url) => {
    send_and_interpret_bot_update(data, update_url, null, orders_request_success_callback, position_request_failure_callback);
    modalElement.unbind("keypress");
    modalElement.modal("hide");
}

const handle_confirm = (modalElement, confirmButton, data, update_url, disable_cancel_buttons) => {
    if (disable_cancel_buttons){
        disable_cancel_all_buttons();
    }
    send_and_interpret_bot_update(data, update_url, null, orders_request_success_callback, orders_request_failure_callback);
    modalElement.unbind("keypress");
    modalElement.modal("hide");
}

const add_cancel_individual_orders_buttons = () => {
    $("button[action=cancel_order]").each((_, element) => {
        $(element).off("click");
        $(element).on("click", (event) => {
            cancel_after_confirm($('#CancelOrderModal'), $(event.currentTarget).attr("order_desc"), $(event.currentTarget).attr(update_url_attr));
        });
    });
}

const disable_cancel_all_buttons = () => {
    $("#cancel_all_orders").prop("disabled",true);
    $("#cancel_order_progress_bar").show();
    const cancelIcon = $("#cancel_all_icon");
    cancelIcon.removeClass("fas fa-ban");
    cancelIcon.addClass("fa fa-spinner fa-spin");
    $("button[action=cancel_order]").each((_, jsElement) => {
        $(jsElement).prop("disabled",true);
    });
}

const orders_request_success_callback = (updated_data, update_url, dom_root_element, msg, status) => {
    if(msg.hasOwnProperty("title")){
        create_alert("success", msg["title"], msg["details"]);
    }else{
        create_alert("success", msg, "");
    }
    debouncedReloadDisplay();
}

const orders_request_failure_callback = (updated_data, update_url, dom_root_element, msg, status) => {
    create_alert("error", msg.responseText, "");
    debouncedReloadDisplay();
}

const position_request_failure_callback = (updated_data, update_url, dom_root_element, msg, status) => {
    create_alert("error", msg.responseText, "");
}

const async_get_data_from_url = async (element) => {
    const url = element.data("url");
    if(typeof url === "undefined"){
        return [];
    }
    return await async_send_and_interpret_bot_update(null, url, null, "GET", true)
}

const reload_positions = async (update) => {
    const table = $("#positions-table");
    const closePositionUrl = table.data("close-url");
    const positions = await async_get_data_from_url(table)
    $("#positions-waiter").hide();
    displayPositionsTable("positions-table", positions, closePositionUrl, update);
}

const reload_trades = async (update) => {
    const table = $("#trades-table");
    const refMarket = table.data("reference-market");
    const trades = await async_get_data_from_url(table)
    $("#trades-waiter").hide();
    return displayTradesTable("trades-table", trades, refMarket, update);
}

const reload_orders = async (update) => {
    const table = $("#orders-table");
    const cancelOrderUrl = table.data("cancel-url");
    const orders = await async_get_data_from_url(table)
    $("#orders-waiter").hide();
    displayOrdersTable("orders-table", orders, cancelOrderUrl, update);
}

const registerScaleSelector = () => {
    $('a[data-action="change-scale"]').on("click", (event) => {
        const selector = $(event.currentTarget);
        if(!selector.hasClass("active")){
            selector.addClass("active");
            $('a[data-action="change-scale"]').each((_, jselement) => {
                const element = $(jselement);
                if(element.data("scale") !== selector.data("scale")){
                    element.removeClass("active");
                }
            })
            reload_pnl(true);
        }
    })
}
const registerSymbolSelector = () => {
    $("#symbol-select").on("change", () => {
        reload_pnl(true);
    })
}
const registerFilterSelectors = () => {
    registerScaleSelector();
    registerSymbolSelector();
}
const getScale = () => {
    return $('a.nav-link.scale-selector.active').data("scale");
}
const getSymbol = () => {
    return $("#symbol-select").val();
}
const updateTradesCount = (pnlHistory) => {
    $("#match-trades-count").text(pnlHistory.reduce((sum, element) => sum + element.tc, 0));
}
const reload_pnl = async (update) => {
    if ($("#pnl_historyChart").length){
        const pnlHistory = await fetchPnlHistory(getScale(), getSymbol());
        loadPnlFullChartHistory(pnlHistory, update);
        loadPnlTableHistory(pnlHistory, update);
        updateTradesCount(pnlHistory);
        $("#pnl-waiter").hide();
    }
}

const resizePnlChart = () => {
    Plotly.Plots.resize("pnl_historyChart")
}

const ordersNotificationCallback = (title, _) => {
    if(title.toLowerCase().indexOf("order") !== -1){
        debouncedReloadDisplay();
    }
}

function debouncedReloadDisplay(){
    debounce(
        () => reloadDisplay(true),
        500
    );
}

const reloadDisplay = async (update) => {
    if(!update){
        await reload_pnl(update);
    }
    await reload_orders(update);
    await reload_positions(update);
    if(await reload_trades(update) && update){
        // only update pnl when a new trade appeared
        await reload_pnl(update);
    }
}

const onPnlTabShow = (e) => {
    resizePnlChart();
}

const registerOnTabShownEvents = () => {
    $("#panel-pnl-tab").on("shown.bs.tab", (e) => onPnlTabShow(e));
}

const registerTableButtonsEvents = () => {
    $("#positions-table").on("draw.dt", () => {
        handle_close_buttons();
    });
    $("#orders-table").on("draw.dt row-reordered", () => {
        if($("#orders-table").data("cancel-url") === undefined){
            return
        }
        add_cancel_individual_orders_buttons();
        const cancelIcon = $("#cancel_all_icon");
        $("#cancel_order_progress_bar").hide();
        if(cancelIcon.hasClass("fa-spinner")){
            cancelIcon.removeClass("fa fa-spinner fa-spin");
            cancelIcon.addClass("fas fa-ban");
        }
        if ($("button[action=cancel_order]").length === 0){
            $("#cancel_all_orders").prop("disabled", true);
        }else{
            $("#cancel_all_orders").prop("disabled", false);
        }
    });
}

const refreshTables = async () => {
    if (!initializedTables){
        await reloadDisplay(true);
        initializedTables = true
    }
}

let initializedTables = false

selectFirstTab();
registerFilterSelectors();
registerTableButtonsEvents();
update_pairs_colors();
$(".watched_element").each((_, element) => {
    $(element).click(addOrRemoveWatchedSymbol);
});
handle_cancel_buttons();
handleClearButton();
register_notification_callback(ordersNotificationCallback);
await reloadDisplay(false);
registerOnTabShownEvents();
handle_rounded_numbers_display();
try {
    if(registerGraphUpdateCallback !== undefined){
        registerGraphUpdateCallback(refreshTables);
    }
} catch (error){
    // nothing to do, registerGraphUpdateCallback doesn't exist
}
});
